File: probes/cargo-git-runtime/verify.py
from common import *
import statistics, random, stat
D=json.loads((O/'discovery.json').read_text());root=Path(D['checkout'])
# Raw committed blobs, not status/diff and never filters; bounded to the measured fixture.
def verify(root, rev=REV, hashes=True):
 head=git(root,'rev-parse','HEAD').stdout.decode().strip()
 if head!=rev: raise ValueError('HEAD differs')
 tree=git(root,'ls-tree','-r','-z',rev).stdout
 rows=[]
 for record in tree.split(b'\0'):
  if not record:continue
  header,path=record.split(b'\t',1);mode,kind,oid=header.split()
  if kind!=b'blob' or mode not in [b'100644',b'100755']:raise ValueError('unsupported tree entry')
  name=path.decode();p=root/name
  for parent in [p,*p.parents]:
   if parent==root:break
   if parent.is_symlink():raise ValueError('symlink')
  meta=p.lstat()
  if not stat.S_ISREG(meta.st_mode):raise ValueError('nonregular file')
  if bool(meta.st_mode&0o111)!=(mode==b'100755'):raise ValueError('mode differs')
  rows.append((name,oid.decode()))
 if len(rows)>100000 or sum((root/n).stat().st_size for n,_ in rows)>512*1024*1024:raise ValueError('allowance')
 stage=git(root,'ls-files','--stage','-z').stdout
 staged=[]
 for r in stage.split(b'\0'):
  if not r:continue
  h,n=r.split(b'\t',1)
  if h.split()[-1]!=b'0':raise ValueError('unmerged index')
  staged.append(n.decode())
 if staged!=[n for n,_ in rows]:raise ValueError('tracked set differs')
 extras=git(root,'ls-files','--others','--exclude-standard','-z').stdout.split(b'\0')
 if any(p not in (b'',b'.cargo-ok') for p in extras):raise ValueError('untracked source')
 if hashes:
  for start in range(0,len(rows),64):
   batch=rows[start:start+64]
   got=git(root,'hash-object','--no-filters','--',*[n for n,_ in batch]).stdout.decode().splitlines()
   if got!=[h for _,h in batch]:raise ValueError('raw blob mismatch')
 return len(rows)

def main():
 count=verify(root)
 assert int(run([T/'raw-verify',root,REV]).stdout)==count
 controls={}
 f=root/'README.md';original=f.read_bytes();before=f.stat()
 try:
  edit=bytearray(original);edit[0]^=1;f.write_bytes(edit);os.utime(f,ns=(before.st_atime_ns,before.st_mtime_ns))
  assert run([T/'raw-verify',root,REV],ok=False).returncode!=0
  p=run(['cargo','metadata','--offline','--locked','--format-version=1','--manifest-path',T/'discovery/Cargo.toml'])
  controls['metadata_accepts_modified_checkout']=p.returncode==0
  try:verify(root);raise AssertionError('missed raw mutation')
  except ValueError as e:controls['restored_mtime_edit']=str(e)
 finally:f.write_bytes(original);os.utime(f,ns=(before.st_atime_ns,before.st_mtime_ns))
 # Clean filters cannot turn wrong bytes into a matching raw object.
 attr=root/'.git/info/attributes';attr.parent.mkdir(exist_ok=True);old=attr.read_bytes() if attr.exists() else None
 git(root,'config','filter.probe.clean','cat /dev/null');attr.write_text('README.md filter=probe\n')
 try:
  f.write_bytes(original+b'\n')
  assert run([T/'raw-verify',root,REV],ok=False).returncode!=0
  try:verify(root);raise AssertionError('missed filtered mutation')
  except ValueError as e:controls['hostile_filter']=str(e)
 finally:
  f.write_bytes(original);os.utime(f,ns=(before.st_atime_ns,before.st_mtime_ns))
  if old is None:attr.unlink()
  else:attr.write_bytes(old)
  git(root,'config','--remove-section','filter.probe')
 extra=root/'probe-extra.rs';extra.write_text('untracked')
 try:
  assert run([T/'raw-verify',root,REV],ok=False).returncode!=0
  try:verify(root);raise AssertionError('missed extra')
  except ValueError as e:controls['untracked']=str(e)
 finally:extra.unlink()
 mode=f.stat().st_mode
 try:
  f.chmod(mode^0o111)
  assert run([T/'raw-verify',root,REV],ok=False).returncode!=0
  try:verify(root);raise AssertionError('missed mode')
  except ValueError as e:controls['mode']=str(e)
 finally:f.chmod(mode)
 assert verify(root)==count
 helper=T/'tool/target/release/rnx-project-assembly-probe'
 p=run([helper,'fingerprint-v2','native',root,'100000','536870912'],ok=False)
 controls['existing_path_fingerprinter_on_cargo_checkout']={'status':p.returncode,'stderr':p.stderr.decode()}
 # CPU pinning applies to this process and children, including Git.
 cpu=min(os.sched_getaffinity(0));os.sched_setaffinity(0,{cpu})
 samples=[]
 functions={
  'coordinates_only':lambda: git(root,'rev-parse','HEAD'),
  'raw_blob_verification':lambda:run([T/'raw-verify',root,REV]),
  'object_fsck':lambda:git(root,'fsck','--full','--no-reflogs'),
  'path_BLAKE3_fingerprint':lambda:run([helper,'fingerprint-v2','native',T/'developer-runtime','100000','536870912']),
 }
 for fn in functions.values():fn()
 for repeat in range(2):
  order=list(functions)*20;random.Random(670001+repeat).shuffle(order)
  for name in order:
   start=time.perf_counter_ns();functions[name]();samples.append({'repeat':repeat,'case':name,'ms':(time.perf_counter_ns()-start)/1e6})
 save('verification.json',{'cpu':cpu,'count':count,'source_bytes':sum((root/n).stat().st_size for n in git(root,'ls-files','-z').stdout.decode().split('\0') if n),'samples':samples,'medians_ms':{k:statistics.median(x['ms'] for x in samples if x['case']==k) for k in functions},'controls':controls,'limits':'Process wall time; raw verification is the compiled Rust probe with Git subprocesses. Not a launch benchmark. fsck is separate, not included in raw verification. Working-tree checks are non-atomic, ignored data/environment remain non-hermetic.'})
 print('Verification passes',flush=True)
if __name__=='__main__':main()

File: probes/cargo-git-runtime/src/verify.rs
//! Isolated raw-object checkout check. No status/diff/clean-filter comparison.
use std::{collections::HashSet, fs, os::unix::fs::PermissionsExt, path::Path, process::Command};
fn git(root: &Path, args: &[&str]) -> Result<Vec<u8>, String> {
    let mut c = Command::new("git");
    for (k, _) in std::env::vars_os() {
        if k.to_string_lossy().starts_with("GIT_") {
            c.env_remove(k);
        }
    }
    c.env("GIT_CONFIG_GLOBAL", "/dev/null")
        .env("GIT_CONFIG_SYSTEM", "/dev/null")
        .env("GIT_CONFIG_NOSYSTEM", "1")
        .env("GIT_ATTR_NOSYSTEM", "1")
        .env("GIT_NO_REPLACE_OBJECTS", "1")
        .env("GIT_OPTIONAL_LOCKS", "0");
    let p = c
        .arg("-C")
        .arg(root)
        .args([
            "-c",
            "core.autocrlf=false",
            "-c",
            "core.fsmonitor=false",
            "-c",
            "core.hooksPath=/dev/null",
            "-c",
            "core.attributesFile=/dev/null",
        ])
        .args(args)
        .output()
        .map_err(|e| e.to_string())?;
    if !p.status.success() {
        return Err(String::from_utf8_lossy(&p.stderr).into_owned());
    }
    Ok(p.stdout)
}
fn verify(root: &Path, rev: &str) -> Result<usize, String> {
    let head = git(root, &["rev-parse", "HEAD"])?;
    if String::from_utf8_lossy(&head).trim() != rev {
        return Err("HEAD differs".into());
    }
    let raw = git(root, &["ls-tree", "-r", "-z", rev])?;
    let mut rows = Vec::new();
    let mut dirs = HashSet::new();
    let mut total = 0u64;
    for row in raw.split(|b| *b == 0).filter(|r| !r.is_empty()) {
        let pos = row
            .iter()
            .position(|b| *b == b'\t')
            .ok_or("bad tree record")?;
        let fields = std::str::from_utf8(&row[..pos])
            .map_err(|e| e.to_string())?
            .split(' ')
            .collect::<Vec<_>>();
        let name = std::str::from_utf8(&row[pos + 1..]).map_err(|e| e.to_string())?;
        if fields.len() != 3 || fields[1] != "blob" || !matches!(fields[0], "100644" | "100755") {
            return Err("unsupported tree entry".into());
        }
        let p = root.join(name);
        for ancestor in p.parent().unwrap().ancestors() {
            if ancestor == root || !dirs.insert(ancestor.to_owned()) {
                break;
            }
            if !fs::symlink_metadata(ancestor)
                .map_err(|e| e.to_string())?
                .is_dir()
            {
                return Err("symlink/non-directory component".into());
            }
        }
        let m = fs::symlink_metadata(&p).map_err(|e| e.to_string())?;
        if !m.is_file() {
            return Err("nonregular file".into());
        }
        if (m.permissions().mode() & 0o111 != 0) != (fields[0] == "100755") {
            return Err("mode differs".into());
        }
        total += m.len();
        if rows.len() >= 100_000 || total > 512 * 1024 * 1024 {
            return Err("allowance".into());
        }
        rows.push((name.to_owned(), fields[2].to_owned()));
    }
    let index = git(root, &["ls-files", "--stage", "-z"])?;
    let mut staged = Vec::new();
    for row in index.split(|b| *b == 0).filter(|r| !r.is_empty()) {
        let pos = row
            .iter()
            .position(|b| *b == b'\t')
            .ok_or("bad index record")?;
        if !row[..pos].ends_with(b" 0") {
            return Err("unmerged index".into());
        }
        staged.push(std::str::from_utf8(&row[pos + 1..]).map_err(|e| e.to_string())?);
    }
    if staged != rows.iter().map(|r| r.0.as_str()).collect::<Vec<_>>() {
        return Err("tracked set differs".into());
    }
    let extra = git(root, &["ls-files", "--others", "--exclude-standard", "-z"])?;
    if extra
        .split(|b| *b == 0)
        .any(|p| !p.is_empty() && p != b".cargo-ok")
    {
        return Err("untracked source".into());
    }
    for chunk in rows.chunks(64) {
        let mut args = vec!["hash-object", "--no-filters", "--"];
        args.extend(chunk.iter().map(|r| r.0.as_str()));
        let hashes = git(root, &args)?;
        if String::from_utf8_lossy(&hashes).lines().collect::<Vec<_>>()
            != chunk.iter().map(|r| r.1.as_str()).collect::<Vec<_>>()
        {
            return Err("raw blob mismatch".into());
        }
    }
    Ok(rows.len())
}
fn main() {
    let args = std::env::args().collect::<Vec<_>>();
    match verify(Path::new(&args[1]), &args[2]) {
        Ok(n) => println!("{n}"),
        Err(e) => {
            eprintln!("{e}");
            std::process::exit(1);
        }
    }
}
